File: src/day2/solution.py
# we want the number of valid passwords, i.e. number of lines that obey the rules

from collections import Counter

# get the two numbers, the character and the string password respectively from the line
def getTokens(line) -> (int, int, str, str):
    tokens = line.split()
    (num1, num2) = map(int, tokens[0].split("-"))
    ch = tokens[1][0] # since we know it's a single char
    psswd = tokens[-1]
    return (num1, num2, ch, psswd)

def isValidPart1(line) -> bool:
    lower, upper, ch, password = getTokens(line)
    return lower <= Counter(password)[ch] <= upper

def isValidPart2(line) -> bool:
    pos_a, pos_b, ch, password = getTokens(line)
    # get the characters of the indexes (modify them to zero based)
    # in the password, check only one of them = ch
    from operator import xor
    return xor((password[pos_a - 1] == ch), (password[pos_b - 1] == ch)) 

if __name__ == "__main__":
    # filter out valid lines and return how many of them are there
    # O(N) time and space, N is the number of passwords
    with open("input.txt") as f:
        print("Part 1: ", len(list(filter(isValidPart1, f))))
    
    with open("input.txt") as f:
        print("Part 2: ", len(list(filter(isValidPart2, f))))

File: src/day2/cpp/solution.cpp
#include <algorithm>
#include <fstream>
#include <string>
#include <tuple>

#include <fmt/core.h>
#include <fmt/ranges.h>
#include <range/v3/all.hpp>

namespace rs = ranges;
namespace rv = ranges::views;

std::tuple<int, int, char, std::string> get_tokens(const std::string &line) {
  auto tokens = line | rv::split(' ') | rs::to<std::vector<std::string>>();
  auto nums = tokens[0] | rv::split('-') | rs::to<std::vector<std::string>>();
  return std::make_tuple(std::stoi(nums[0]), std::stoi(nums[1]), tokens[1][0],
                         tokens.back());
}

std::vector<std::string> parse(std::istream &&input) {
  return rs::getlines(input) | rs::to_vector;
}

auto part1(const std::vector<std::string> &input) {
  return rs::count_if(input, [](const auto &line) {
    auto [lower, upper, ch, password] = get_tokens(line);
    auto freq = rs::count(password, ch);
    return freq >= lower && freq <= upper;
  });
}

auto part2(const std::vector<std::string> &input) {
  return rs::count_if(input, [](const auto &line) {
    auto [pos_a, pos_b, ch, password] = get_tokens(line);
    return (password[pos_a - 1] == ch) ^ (password[pos_b - 1] == ch);
  });
}

int main() {
  auto input = parse(std::ifstream{"../../input.txt", std::ios::in});
  fmt::print("Part 1: {}\n", part1(input));
  fmt::print("Part 2: {}\n", part2(input));
  return 0;
}